cpp/src/buffer/buffer.cpp

```diff
@@ -50,17 +68,22 @@ std::unique_ptr<Buffer> Buffer::copy(rmm::cuda_stream_view stream) const {
     return std::visit(
         overloaded{
             [&](const HostStorageT& storage) -> std::unique_ptr<Buffer> {
-                return std::unique_ptr<Buffer>(
-                    new Buffer{std::make_unique<std::vector<uint8_t>>(*storage), br}
-                );
-            },
-            [&](const DeviceStorageT& storage) -> std::unique_ptr<Buffer> {
+                // Have to ensure that any async work is complete.
+                synchronize();
```

Comment:
Why do you have to ensure async work is complete? Your source is host memory, seems like anything that has been done with it should have been synchronized before, it should never be up to the consumer to magically figure out how the host buffer has been processed, how do you know it is even the right stream then?

Reply:
Suppose you spill a device side buffer to `Buffer` as host memory. You did that with `cudaMemcpyAsync`. Now you have a `Buffer` with host memory, but it might not be `ready`. So if you then copy it to a new `Buffer` with host memory you can either fail if it's not `is_ready` or you synchronize.

Reply:
Yeah, ok, but when would that happen in practice? I'd argue the consumer needs to checking the `Buffer` (or whatever other object) that delivered the host memory should be checked if `is_ready()`, otherwise, to use your words, we are now putting a "big hammer" (although slightly smaller than device sync) here. What is the argument for whoever called `Buffer::copy()` not to have ensured that in the first place?

Reply:
If we don't do the synchronisation then every time you do a buffer host-to-host copy you must check that that is what you're doing and then `sync`. Or else check `is_ready` and then if you're not ready, push it back into a copy queue?

That seems like a very difficult to manage API.

Reply:
In almost all cases the source host buffer will have no event and thus the sync is a noop. But sometimes it might